lasagne/tests/layers/test_autoenc.py

```diff
@@ -0,0 +1,159 @@
+            incoming=layer,
+            layer=layer
+        )
+
+        return {
+            'layer': layer,
+            'invlayer': invlayer,
+        }
+
+    def test_init(self, invlayer_vars):
+        layer = invlayer_vars['layer']
+        invlayer = invlayer_vars['invlayer']
+        # Check that the output shape of the invlayer is the same
+        # as the input shape of the layer
+        assert layer.input_shape == invlayer.output_shape
+
+    def test_get_output_shape_for(self, invlayer_vars):
+        invlayer = invlayer_vars['invlayer']
+        assert invlayer.get_output_shape_for(
+            [(34, 55, 89, 144), (5, 8, 13, 21), (1, 1, 2, 3)]) == (1, 1, 2, 3)
+
+    def test_get_output_for(self, invlayer_vars):
+        from lasagne.layers.helper import get_output
+        invlayer = invlayer_vars['invlayer']
+        layer = invlayer_vars['layer']
+        W = layer.W.get_value()
+        input = theano.shared(
+            np.random.rand(*layer.input_shape))
+        results = get_output(invlayer, inputs=input)
+
+        # Check that the output of the invlayer is the output of the
+        # dot product of the output of the dense layer and the
+        # transposed weights
+        assert np.allclose(
+            results.eval(), np.dot(np.dot(input.get_value(), W), W.T))
+
+
+def test_build_autoencoder():
+
+    from lasagne.layers import InputLayer, DenseLayer, BiasLayer
+    from lasagne.layers import InverseLayer, NonlinearityLayer
+    from lasagne.layers import build_autoencoder
+    from lasagne.nonlinearities import identity
+    from lasagne.layers import get_all_layers, get_output
+
+    input_shape = tuple(np.random.randint(20, high=100, size=2))
+    l_in = InputLayer(input_shape)
+    l1 = DenseLayer(l_in, num_units=50,
+                    nonlinearity=identity,
+                    b=None)
+    l2 = DenseLayer(l1, num_units=10,
+                    nonlinearity=identity,
+                    b=None)
+    l_ae = build_autoencoder(l2,
+                             nonlinearity='same',
+                             b=None)
+
+    # Check that all layers are the right type
+    autoencoder_layers = get_all_layers(l_ae)
+    assert isinstance(autoencoder_layers[0], InputLayer)
+    assert isinstance(autoencoder_layers[1], DenseLayer)
+    assert isinstance(autoencoder_layers[2], BiasLayer)
+    assert isinstance(autoencoder_layers[3], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[4], DenseLayer)
+    assert isinstance(autoencoder_layers[5], BiasLayer)
+    assert isinstance(autoencoder_layers[6], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[7], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[8], BiasLayer)
+    assert isinstance(autoencoder_layers[9], InverseLayer)
+    assert isinstance(autoencoder_layers[10], NonlinearityLayer)
+    assert isinstance(autoencoder_layers[11], BiasLayer)
+    assert isinstance(autoencoder_layers[12], InverseLayer)
```

Comment:
Wait, that's not what I would expect the decoder to look like. It should be `InverseLayer` -> `BiasLayer` -> `NonlinearityLayer` -> `InverseLayer` -> `BiasLayer` -> `NonlinearityLayer`. Maybe you should re-think splitting up the creation of the auto-encoder into two functions, and deal with everything in one go. Otherwise `build_autoencoder` needs to deal with reordering things correctly.